design-docs/012-python-namespaces

```diff
@@ -0,0 +1,216 @@
+- This proposal chooses to group by related functionality.
+- The location of classes that are only really returned to the user, and rarely directly called by the user (e.g. `Diff`), doesn't really matter.
+  - Similarly classes with no constructor methods (e.g. `Session`) don't need to be in the top-level namespace.
+- Modules with only one member are fine if they add clarity.
+- Adding an extra intermediate namespace doesn't make things much harder for users to find. In other words this:
+
+```python
+import icechunk
+storage = icechunk.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+is not significantly harder than this:
+
+```python
+import icechunk
+storage = icechunk.storage.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+Also note that this is already valid API:
+
+```python
+import icechunk
+storage = icechunk.Storage.s3_storage(bucket="my-bucket", prefix="my-prefix", from_env=True)
+repo = icechunk.Repository.create(storage)
+```
+
+### Top-level `icechunk` module:
+
+- `icechunk.Repository` (also available in `icechunk.repository.Repository`)
+- `icechunk.IcechunkStore` (also available in `icechunk.store.IcechunkStore`)
+- `icechunk.spec_version`
+- `icechunk.print_debug_info`
+- `icechunk.__version__`
+
+### Sub-modules
+
+Existing namespaces, with new members added:
+
+- `credentials`
+  - `credentials.Credentials`
+  - `credentials.azure_credentials`
+  - `credentials.azure_from_env_credentials`
+  - `credentials.azure_static_credentials`
+  - `credentials.containers_credentials`
+  - `credentials.gcs_credentials`
+  - `credentials.gcs_from_env_credentials`
+  - `credentials.gcs_refreshable_credentials`
+  - `credentials.gcs_static_credentials`
+  - `credentials.s3_anonymous_credentials`
+  - `credentials.s3_credentials`
+  - `credentials.s3_from_env_credentials`
+  - `credentials.s3_refreshable_credentials`
+  - `credentials.s3_static_credentials`
+- `dask`
+  - `dask.computing_meta`
+  - `dask.store_dask`
+- `distributed`
+  - `distributed.extract_sessions`
+  - `distributed.merge_sessions`
+- `garbage`
+  - `garbage.GCSummary`
+- `repository`
+  - `repository.Repository`
+- `session`
+  - `session.Session`
+  - `session.ForkSession`
+- `storage`
+  - `storage.S3Options`
+  - `storage.Storage`
+  - `storage.StorageConcurrencySettings`
+  - `storage.StorageRetriesSettings`
+  - `storage.StorageSettings`
+  - `storage.AnyObjectStoreConfig`
+  - `storage.azure_storage`
+  - `storage.gcs_storage`
+  - `storage.gcs_store`
+  - `storage.http_store`
+  - `storage.in_memory_storage`
+  - `storage.local_filesystem_storage`
+  - `storage.local_filesystem_store`
+  - `storage.r2_storage`
+  - `storage.s3_storage`
+  - `storage.s3_store`
+  - `storage.tigris_storage`
+- `store`
+  - `store.IcechunkStore`
+- `xarray`
+  - `xarray.to_icechunk`
```

Comment:
I wonder if `to_icechunk` should be at the top level.

Reply:
We might want to leave open the possibility of adding API for serializing non-xarray things `to_icechunk` though...